### ISSUE TITLE
[uri module] TLS SNI support missing?

### ISSUE THREAD

**Issue by fdutheil** [ansible/ansible-modules-core#3355]
##### ISSUE TYPE

- Bug Report
##### COMPONENT NAME

URI module
##### ANSIBLE VERSION

```
ansible 2.0.1.0
  config file = /home/florent/.ansible.cfg
  configured module search path = Default w/o overrides
```
##### CONFIGURATION

Nothing fancy, ssh pipelining and control/master feature enabled in .ansible.cfg.
##### OS / ENVIRONMENT

Ansible running from Ubuntu 14.04, managing CentOS 6.7, but task running locally (delegated). The test case provided bellow entirely runs locally.
##### SUMMARY

TLS connection fails when using the uri module and targeting a webserver/platform using TLS SNI.
The last time I had troubles in the very same conditions and targeting the very same environment was because the stress-test tools didn't support (or weren't configured by default to support) TLS SNI.
##### STEPS TO REPRODUCE

Use the following playbook:

```

---
- name: SNI PLaybook test
  hosts: localhost
  gather_facts: no
  tasks:
    - name: HTTP GET
      uri:
        url: "{{ url }}"
        method: GET
```

And use an URL toward a TLS SNI enabled target, for example (real URL substituted):
`ansible-playbook test.yml -i inventory -e "url=https://myvirtualhostedwebserver.com/doku.php"`

##### EXPECTED RESULTS

The task should complete normally, getting a TLS connection up and getting an HTTP code 200. 

```
$ ansible-playbook test.yml -i inventory -e "url=https://www.google.com/"

PLAY [SNI PLaybook test] *******************************************************

TASK [HTTP GET] ****************************************************************
ok: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0
```
##### ACTUAL RESULTS

```
$ ansible-playbook test.yml -i inventory -e "url=https://myvirtualhostedwebserver.com/doku.php" -vvv
Using /home/florent/.ansible.cfg as config file
1 plays in test.yml

PLAY [SNI PLaybook test] *******************************************************

TASK [HTTP GET] ****************************************************************
task path: /home/florent/Documents/Work/Ysance/Projets/Anaplan/GIT/magento-anaplan/ansible/test.yml:6
ESTABLISH LOCAL CONNECTION FOR USER: florent
127.0.0.1 EXEC /bin/sh -c 'mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1459503447.73-273175456065301 `" && echo "` echo $HOME/.ansible/tmp/ansible-tmp-1459503447.73-273175456065301 `"'
127.0.0.1 PUT /tmp/tmpJuOxhw TO /home/florent/.ansible/tmp/ansible-tmp-1459503447.73-273175456065301/uri
127.0.0.1 EXEC /bin/sh -c 'LANG=fr_FR.UTF-8 LC_ALL=fr_FR.UTF-8 LC_MESSAGES=fr_FR.UTF-8 /usr/bin/python /home/florent/.ansible/tmp/ansible-tmp-1459503447.73-273175456065301/uri; rm -rf "/home/florent/.ansible/tmp/ansible-tmp-1459503447.73-273175456065301/" > /dev/null 2>&1'
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_args": {"backup": null, "body": null, "body_format": "raw", "content": null, "creates": null, "delimiter": null, "dest": null, "directory_mode": null, "follow": false, "follow_redirects": "safe", "force": null, "force_basic_auth": false, "group": null, "method": "GET", "mode": null, "owner": null, "password": null, "regexp": null, "remote_src": null, "removes": null, "return_content": false, "selevel": null, "serole": null, "setype": null, "seuser": null, "src": null, "status_code": [200], "timeout": 30, "url": "https://myvirtualhostedwebserver.com/doku.php", "user": null, "validate_certs": true}, "module_name": "uri"}, "msg": "Unable to validate server's certificate against available CA certs."}

NO MORE HOSTS LEFT *************************************************************
    to retry, use: --limit @/home/florent/.ansible/retry/test.retry

PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1 
```

See bellow a successful curl command on the same environment. Of course, the TLS certificate is valid (no custom CA).

```
$ curl -I https://myvirtualhostedwebserver.com/doku.php
HTTP/1.1 200 OK
Server: nginx
Date: Fri, 01 Apr 2016 09:19:31 GMT
Content-Type: text/html; charset=utf-8
Connection: keep-alive
Keep-Alive: timeout=20
Set-Cookie: DokuWiki=s9p7koh31c1e638sg2vpvhmv05; path=/; secure; HttpOnly
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Pragma: no-cache
Set-Cookie: DW7fa065a06cb74b536c124cfbe56ac6d3=deleted; expires=Thu, 01-Jan-1970 00:00:01 GMT; Max-Age=0; path=/; secure; httponly
X-UA-Compatible: IE=edge,chrome=1
Strict-Transport-Security: max-age=31536000; includeSubDomains
X-Frame-Options: DENY
X-Content-Type-Options: nosniff
```

Using `validate_certs: no` is a workaround, but is not desirable.

**Comment by sivel** [comment]
I believe this may already be fixed in both devel and 2.0.2 (which just had an RC released).

**Comment by tumbl3w33d** [comment]
The same thing happened to me today using the rpm_key module. Not sure if I should create a  separate issue for it.
I tried to add the gpg key from https://yum.dockerproject.org/gpg which is obviously behind cloudfront now. This was working fine before, until today it started failing, saying

> SSL Certificate does not belong to yum.dockerproject.org.  Make sure the url has a certificate that belongs to it or use validate_certs=False (insecure)

so I tried it in the browser and the certificate seemed fine. It made me realize it might be a SNI issue and right, if you check:
`openssl s_client -showcerts -connect yum.dockerproject.org:443`
you get a cert for CN=res.ebdcdn.com while for
`openssl s_client -showcerts -servername yum.dockerproject.org -connect yum.dockerproject.org:443`
you get a cert for CN=*.dockerproject.org as you would expect.

validate_certs=no might help out for now, but it's ugly if you use a third party role (like I use jtyr.yumrepo) which has certificate validation active.

ansible 2.0.1.0

**Comment by artbw** [comment]
I did some research and wanna share what I found. I found a test site for SNI and forged a request to it:
>import urllib2
>response = urllib2.urlopen('https://sni.velox.ch/')
>html = response.read()
>print(html)

The response was:

> <p><strong>Unfortunately, your client </strong>[Python-urllib/2.6] <strong> 
did not send a TLS server name indication extension

So I did more research about SNI and python and found out, that it's not supported before python 2.7. At least that's what I understand from:

- http://bugs.python.org/issue5639
- http://stackoverflow.com/questions/19477214/urllib3-on-python-2-7-sni-error-on-google-app-engine/19477363#19477363

Now I would like to know, what that means for ansible. I guess you use urllib2 for quite a few modules and all of them are affected by this limitation.  How are we supposed to deal with this, if we are using, for example CentOS release 6.7 (Final) which comes with Python 2.6.6?

**Comment by tumbl3w33d** [comment]
I did some research and wanna share what I found. I found a test site for SNI and forged a request to it:

> import urllib2
> response = urllib2.urlopen('https://sni.velox.ch/')
> html = response.read()
> print(html)

The response was:

> Unfortunately, your client[Python-urllib/2.6] 
> did not send a TLS server name indication extension

So I did more research about SNI and python and found out, that it's not supported before python 2.7. At least that's what I understand from:
- http://bugs.python.org/issue5639
- http://stackoverflow.com/questions/19477214/urllib3-on-python-2-7-sni-error-on-google-app-engine/19477363#19477363

Now I would like to know, what that means for ansible. I guess you use urllib2 for quite a few modules and all of them are affected by this limitation.  How are we supposed to deal with this, if we are using, for example CentOS release 6.7 (Final) which comes with Python 2.6.6?

**Comment by sivel** [comment]
SNI support requires python>=2.7.9.  There is an error string in urls.py that lists this as a requirement.

Our only option to potentially support SNI in earlier versions would be to implement something like what urllib3 does, by requiring pyopenssl, ndg-httpsclient, and pyasn1.

The implementations of what are used in ansible currently and urllib3, do not appear to be immediately compatible.

**Comment by sivel** [comment]
Linking what I am referring to in urllib3:  https://github.com/shazow/urllib3/blob/master/urllib3/contrib/pyopenssl.py

Again, I'm not sure in the feasibility of integrating that implementation, but we may be able to look into it.

**Comment by sivel** [comment]
Here is a super early proof of concept that integrates urllib3 pyopenssl into urls.py:

https://github.com/sivel/ansible/commit/c0c284e8519f738342213e8a80fd7b1d8a01949a

It does additionally require:

```
pip install pyopenssl ndg-httpsclient pyasn1
```

This does seem to be working for me in Python 2.6.9.  I have not tried elsewhere really.

**Comment by ghost** [comment]
Good job. I can confirm it's working on python 2.6.6. Thank you very much :)

**Comment by sivel** [comment]
I've opened a PR which is a bit more simple at https://github.com/ansible/ansible/pull/15289 for supporting SNI on Py26 or older Py2.7<2.7.9

The change is that there is a dependency on urllib3 also.

**Comment by sivel** [comment]
The above referenced PR has been merged into devel, which will become v2.1.
